packages/docusaurus/src/commands/swizzle.ts

```diff
@@ -17,26 +18,76 @@ export async function swizzle(
   siteDir: string,
   themeName: string,
   componentName?: string,
+  typescript?: boolean,
 ): Promise<void> {
   const plugin: any = importFresh(themeName);
   const context = loadContext(siteDir);
   const pluginInstance = plugin(context);
-  let fromPath = pluginInstance.getThemePath();
+  const themePath = pluginInstance.getThemePath();
+  const libThemePath = path.join(themePath, '..', '..', 'lib', 'theme');
+  let fromPath = themePath;
 
   if (fromPath) {
-    let toPath = path.resolve(siteDir, THEME_PATH);
+    const toThemePath = path.resolve(siteDir, THEME_PATH);
+    let toPath = toThemePath;
     if (componentName) {
       fromPath = path.join(fromPath, componentName);
       toPath = path.join(toPath, componentName);
 
-      // Handle single JavaScript file only.
-      // E.g: if <fromPath> does not exist, we try to swizzle <fromPath>.js instead
-      if (!fs.existsSync(fromPath) && fs.existsSync(`${fromPath}.js`)) {
-        [fromPath, toPath] = [`${fromPath}.js`, `${toPath}.js`];
+      // Handle single TypeScript/JavaScript file only.
+      // E.g: if <fromPath> does not exist, we try to swizzle <fromPath>.(ts|tsx|js) instead
+      if (!fs.existsSync(fromPath)) {
+        if (fs.existsSync(`${fromPath}.ts`)) {
+          [fromPath, toPath] = [`${fromPath}.ts`, `${toPath}.ts`];
+        } else if (fs.existsSync(`${fromPath}.tsx`)) {
+          [fromPath, toPath] = [`${fromPath}.tsx`, `${toPath}.tsx`];
+        } else if (fs.existsSync(`${fromPath}.js`)) {
+          [fromPath, toPath] = [`${fromPath}.js`, `${toPath}.js`];
+        }
       }
     }
     await fs.copy(fromPath, toPath);
 
+    // If the user doesn't want swizzle TypeScript components, we replace the swizzled TypeScript
+    // component by traspiled JavaScript.
+    if (!typescript) {
```

Comment:
Actually we have a problem here.
Suppose a third-party theme ships TS theme components by default, and don't provide a JS fallback at `lib/theme`, then `swizzle` would crash here.

An easy but dirty fix is to guard this block by
```ts
if (themeName === '@docusaurus/theme-classic') {
 // ...
}
```
but this shouldn't be a long-term solution.

Everything would be much cleaner if we can make emitting TS components as the default behavior, and emit JS components as fallback if user passes some flag like `--prefer-js`. However, this does require us to change the default emit to TypeScript, which might break some existing setup.

I would like to ask ideas and suggestions from maintainers before I dig deeper into this. @yangshun @lex111

Reply:
Outdated